ext/sockets/sockets.c

```diff
@@ -1413,6 +1435,54 @@ PHP_FUNCTION(socket_bind)
 }
 /* }}} */
 
+#ifdef AF_PACKET
+#define ETH_SUB_CHECKLENGTH(a, lyr)						\
+	do {									\
+		if ((char *)ipdata + sizeof(a) < ZSTR_VAL(recv_buf) + slen) {	\
+			zend_string_efree(recv_buf);				\
+			Z_DELREF_P(zpayload);					\
+			ZEND_TRY_ASSIGN_REF_VALUE(arg2, obj);			\
+			ZEND_TRY_ASSIGN_REF_STRING(arg5, ifrname);		\
+			if (arg6) {						\
+				ZEND_TRY_ASSIGN_REF_LONG(arg6, sll.sll_ifindex);\
+			}							\
+			zend_value_error("invalid %s header", lyr);		\
+			return FAILURE;						\
+		}								\
+	} while (0)
+static zend_result php_socket_afpacket_add_tcp(unsigned char *ipdata, struct sockaddr_ll sll, char *ifrname, zend_string *recv_buf,
+					       size_t slen, zval *szpayload,  zval *zpayload, zval *obj, zval *arg2, zval *arg5, zval *arg6) {
+	struct tcphdr a;
+	ETH_SUB_CHECKLENGTH(a, "TCP");
+	memcpy(&a, ipdata, sizeof(a));
+	struct tcphdr *tcp = &a;
+	object_init_ex(szpayload, tcppacket_ce);
+	zend_update_property_long(Z_OBJCE_P(szpayload), Z_OBJ_P(szpayload), ZEND_STRL("srcPort"), ntohs(tcp->th_sport));
+	zend_update_property_long(Z_OBJCE_P(szpayload), Z_OBJ_P(szpayload), ZEND_STRL("dstPort"), ntohs(tcp->th_dport));
+	zend_update_property_long(Z_OBJCE_P(szpayload), Z_OBJ_P(szpayload), ZEND_STRL("headerSize"), sizeof(*tcp));
+	zend_update_property_string(Z_OBJCE_P(szpayload), Z_OBJ_P(szpayload), ZEND_STRL("rawPacket"), (char *)ipdata);
```

Comment:
We should use `zend_update_property_stringl()` here (and in other places), as `ipdata` is not a null-terminated string